src/NHibernate/Criterion/InExpression.cs

```diff
@@ -62,41 +58,50 @@ public override SqlString ToSqlString(ICriteria criteria, ICriteriaQuery criteri
 				return new SqlString("1=0");
 			}
 
-			//TODO: add default capacity
-			SqlStringBuilder result = new SqlStringBuilder();
-			SqlString[] columnNames =
-				CriterionUtil.GetColumnNames(_propertyName, _projection, criteriaQuery, criteria);
-			
-			// Generate SqlString of the form:
-			// columnName1 in (values) and columnName2 in (values) and ...
-			Parameter[] parameters = GetParameterTypedValues(criteria, criteriaQuery).SelectMany(t => criteriaQuery.NewQueryParameter(t)).ToArray();
+			SqlString[] columns = CriterionUtil.GetColumnNames(_propertyName, _projection, criteriaQuery, criteria);
 
-			for (int columnIndex = 0; columnIndex < columnNames.Length; columnIndex++)
+			var list = new List<Parameter>(columns.Length * Values.Length);
+			foreach (var typedValue in GetParameterTypedValues(criteria, criteriaQuery))
 			{
-				SqlString columnName = columnNames[columnIndex];
-
-				if (columnIndex > 0)
-				{
-					result.Add(" and ");
-				}
+				//Must be executed after CriterionUtil.GetColumnNames (as it might add _projection parameters to criteria)
+				list.AddRange(criteriaQuery.NewQueryParameter(typedValue));
+			}
 
-				result
-					.Add(columnName)
-					.Add(" in (");
+			var bogusParam = Parameter.Placeholder;
 
-				for (int i = 0; i < _values.Length; i++)
-				{
-					if (i > 0)
-					{
-						result.Add(StringHelper.CommaSpace);
-					}
-					result.Add(parameters[i]);
-				}
+			var sqlString = GetSqlString(criteriaQuery, columns, bogusParam);
+			sqlString.SubstituteBogusParameters(list, bogusParam);
+			return sqlString;
+		}
 
-				result.Add(")");
+		private SqlString GetSqlString(ICriteriaQuery criteriaQuery, SqlString[] columns, Parameter bogusParam)
+		{
+			if (columns.Length <= 1 || criteriaQuery.Factory.Dialect.SupportsRowValueConstructorSyntaxInInList)
+			{
+				var parens = columns.Length > 1 ? new[] {new SqlString("("), new SqlString(")"),} : null;
+				SqlString comaSeparator = new SqlString(", ");
+				var singleValueParam = SqlStringHelper.Repeat(new SqlString(bogusParam), columns.Length, comaSeparator, parens);
+
+				var parameters = SqlStringHelper.Repeat(singleValueParam, Values.Length, comaSeparator, null);
+
+				//single column: col1 in (?, ?)
+				//multi column:  (col1, col2) in ((?, ?), (?, ?))
+				return new SqlString(
+					parens?[0] ?? SqlString.Empty,
```

Comment:
`parens` is either array containing open and closed parens `new[]{"(", ")"}` (only it's `SqlString` not string) or null if no parens are required.
So this is equivalent of:
```C#
bool needWrapInParens = columns.Length > 1;
return new SqlString(
	needWrapInParens ? new SqlString("(") : SqlString.Empty, //instead of parens?[0] ?? SqlString.Empty,
...
	needWrapInParens ? new SqlString(")") : SqlString.Empty,//instead of parens?[0] ?? SqlString.Empty,
```

Only those `parens` are also reused in `SqlStringHelper.Repeat`